### PR TITLE
stats: modernize stats collection and implement Pod level stats reporting

### PR DESCRIPTION
**Pull request by haircommander**
#### What type of PR is this?

/kind feature
#### What this PR does / why we need it:
CRI stats has been in sore need of an update, and is targeted to be updated in [1.23](https://github.com/kubernetes/enhancements/issues/2371). Part of this update is support in the main CRI implementations.

This PR is partially a refactor and partially a feature. In includes cleaning up some code that wasn't actually used and was thus wasteful, bumping CRI to get the new pod level stats calls, adding a StatsServer to give the ability to cache the stats on a periodic basis (rather than collect on-demand, as is default). 

There still should be some test updates, but integration tests will be hard before critools adds support for the new fields and cuts a release. In the meantime, lots of this is reviewable.
#### Does this PR introduce a user-facing change?

```release-note
Add a [crio.stats] table to the config for configuring all things stats-related.
Add ability (with config field StatsCollectionPeriod) to calculate and save stats for pods and containers, rather than return them on demand.
Add support for {,List}PodSandboxStats, which allows the kubelet to take advantage of the PodAndContainerStatsFromCRI feature gate--one that is set to be alpha in 1.23.
```

### CODE REVIEW

File: internal/config/cgmgr/stats.go
Comment (saschagrunert) [inline internal/config/cgmgr/stats.go:31]:
```suggestion
	return err
```

File: internal/config/cgmgr/stats.go
Comment (saschagrunert) [inline internal/config/cgmgr/stats.go:46]:
```suggestion
	return err
```

File: internal/config/cgmgr/stats.go
Comment (saschagrunert) [inline internal/config/cgmgr/stats.go:161]: nit
```suggestion
			"Unable to account working set stats: total_inactive_file (%d) > memory usage (%d)",
```

**Review by saschagrunert — APPROVED** [review]
Just a few nits, otherwise LGTM

**Comment by fidencio** [comment]
@haircommander, can you hold this one till Tomorrow?  I'd like to do a review (together with some local tests) on this later Today / Tomorrow morning.

File: internal/oci/runtime_vm.go
Comment (fidencio) [inline internal/oci/runtime_vm.go:811]: For kata-containers, this is the part that differs from the approach taken before and that we should ensure the CRI runtime has access to the correct information.

Those are collect, and then passed up to the CRI runtime, here:
https://github.com/kata-containers/kata-containers/blob/e89c06e68bcd465cc79dc68945751a706f230bfe/src/runtime/pkg/containerd-shim-v2/metrics.go#L98-L146

It means that, theoretically, from the kata-containers everything needed here is already provided by kata-containers, but I still need to get to do some more tests on this.

File: internal/oci/runtime_vm.go
Comment (fidencio) [inline internal/oci/runtime_vm.go:836]: @haircommander, I feel like our future selves would appreciate a comment, probably in the commit message, explaining why CPU*, Net*, and Block* stats coming from the containers are not relevant anymore.

File: server/cri/v1/rpc_list_pod_sandbox_stats.go
Comment (kolyshkin) [inline server/cri/v1/rpc_list_pod_sandbox_stats.go:13]: Seems this should belong to the next commit (the third one in the series, named "stub {List}PodSandboxStats" or so).

Comment (kolyshkin) [reply server/cri/v1/rpc_list_pod_sandbox_stats.go:13]: or maybe not... I'm not quite following the logic

File: internal/oci/oci_linux.go
Comment (kolyshkin) [inline internal/oci/oci_linux.go:1]: It would be nice to have a separate PR that adds go:build tags for all files at once, to minimize the noise introduced by it.

Also, in this particular case, the `linux` build tag is not needed as the file ends with `_linux.go`

So, something similar to https://github.com/opencontainers/runc/pull/3185 would be nice to have.

File: internal/oci/oci_linux.go
Comment (kolyshkin) [inline internal/oci/oci_linux.go]: Maybe use `strconv.ParseUint64` so we don't have to convert.

File: internal/config/cgmgr/cgmgr.go
Comment (kolyshkin) [inline internal/config/cgmgr/cgmgr.go:62]: You can make it more self-documenting here by adding parameter names, e.g.
```go
PopulateContainerCgroupStats(sbParent, cid string, stats *types.ContainerStats) error
```

File: internal/config/cgmgr/cgmgr_test.go
Comment (kolyshkin) [inline internal/config/cgmgr/cgmgr_test.go:20]: is this hunk related?

File: pkg/config/template.go
Comment (kolyshkin) [inline pkg/config/template.go:86]: Can simplify this to use iota. https://github.com/cri-o/cri-o/pull/5469

File: internal/lib/stats/stats_server.go
Comment (kolyshkin) [inline internal/lib/stats/stats_server.go:173]: nit: `logrus.Error(err)`.

File: internal/lib/stats/stats_server.go
Comment (kolyshkin) [inline internal/lib/stats/stats_server.go:187]: Not for this PR, but in general: at some point we should ditch pkg/errors and switch to go native error wrapping (i.e. `fmt.Errorf(".... %w", err)`.

File: internal/lib/container_server.go
Comment (kolyshkin) [inline internal/lib/container_server.go:142]: Why not 
```go
return &ContainerServer{
...
	StatsServer: statsserver.New(c)
...
}
```

Comment (kolyshkin) [reply internal/lib/container_server.go:142]: Ah, because `c`. Nevermind :)

File: server/sandbox_list.go
Comment (kolyshkin) [inline server/sandbox_list.go:97]: Why was this moved? Asking because it makes the review harder (and also git log / git blame).

**Comment by kolyshkin** [comment]
LGTM overall. It's kind of hard to review because of so many code moved around, so I might have missed something.

**Review by kolyshkin — APPROVED** [review]
LGTM

File: internal/oci/oci_linux.go
Comment (haircommander) [reply internal/oci/oci_linux.go:1]: hm this was added by vim-go plugin.. I agree we should probably do it all in one fell swoop, though I may push for this to stay if the plugin continues to add it :sweat_smile:

File: internal/oci/runtime_vm.go
Comment (haircommander) [reply internal/oci/runtime_vm.go:836]: the thing is they have never been relevant. I think they were added because they did come from the cgroup package we use but it does feel odd to leave a comment about code that was dead and wasteful

File: internal/config/cgmgr/cgmgr_test.go
Comment (haircommander) [reply internal/config/cgmgr/cgmgr_test.go:20]: not technically related but also trivial and self evident so I didn't bother adding a different commit

File: pkg/config/template.go
Comment (haircommander) [reply pkg/config/template.go:86]: I've rebased on the pkg/config commit

File: server/sandbox_list.go
Comment (haircommander) [reply server/sandbox_list.go:97]: it is a pet peeve of mine when functions are defined before they're called--preventing one reading the file top-down to understand the flow. I figured this one was a relatively small function and not too touch to check for differences

File: internal/lib/stats/stats_server.go
Comment (haircommander) [reply internal/lib/stats/stats_server.go:187]: agreed!

File: internal/config/cgmgr/cgmgr.go
Comment (haircommander) [reply internal/config/cgmgr/cgmgr.go]: good idea, done

**Comment by haircommander** [comment]
comments addressed, PTAL @cri-o/cri-o-maintainers

**Comment by saschagrunert** [comment]
Unfortunately this needs a rebase.

**Comment by haircommander** [comment]
rebased!

**Review by saschagrunert — APPROVED** [review]
/lgtm

**Pull request merged**
